access/src/main/java/org/apache/accumulo/access/AccessExpressionImpl.java

```diff
@@ -0,0 +1,635 @@
+      }
+      return new BytesWrapper(expression, start, end - start);
+    }
+  }
+
+  /**
+   * A node comparator. Nodes sort according to node type, terms sort lexicographically. AND and OR
+   * nodes sort by number of children, or if the same by corresponding children.
+   */
+  static class NodeComparator implements Comparator<Node>, Serializable {
+
+    private static final long serialVersionUID = 1L;
+    byte[] text;
+
+    /**
+     * Creates a new comparator.
+     *
+     * @param text expression string, encoded in UTF-8
+     */
+    public NodeComparator(byte[] text) {
+      this.text = text;
+    }
+
+    @Override
+    public int compare(Node a, Node b) {
+      int diff = a.type.ordinal() - b.type.ordinal();
+      if (diff != 0) {
+        return diff;
+      }
+      switch (a.type) {
+        case EMPTY:
+          return 0; // All empty nodes are the same
+        case TERM:
+          return Arrays.compare(text, a.start, a.end, text, b.start, b.end);
+        case OR:
+        case AND:
+          diff = a.children.size() - b.children.size();
+          if (diff != 0) {
+            return diff;
+          }
+          for (int i = 0; i < a.children.size(); i++) {
+            diff = compare(a.children.get(i), b.children.get(i));
+            if (diff != 0) {
+              return diff;
+            }
+          }
+      }
+      return 0;
+    }
+  }
+
+  /*
+   * Convenience method that delegates to normalize with a new NodeComparator constructed using the
+   * supplied expression.
+   */
+  public static Node normalize(Node root, byte[] expression) {
+    return normalize(root, new NodeComparator(expression));
+  }
+
+  // @formatter:off
+    /*
+     * Walks an expression's AST in order to:
+     *  1) roll up expressions with the same operant (`a&(b&c) becomes a&b&c`)
+     *  2) sort labels lexicographically (permutations of `a&b&c` are re-ordered to appear as `a&b&c`)
+     *  3) dedupes labels (`a&b&a` becomes `a&b`)
+     */
+    // @formatter:on
+  public static Node normalize(Node root, NodeComparator comparator) {
+    if (root.type != NodeType.TERM) {
+      TreeSet<Node> rolledUp = new TreeSet<>(comparator);
+      java.util.Iterator<Node> itr = root.children.iterator();
+      while (itr.hasNext()) {
+        Node c = normalize(itr.next(), comparator);
+        if (c.type == root.type) {
+          rolledUp.addAll(c.children);
+          itr.remove();
+        }
+      }
+      rolledUp.addAll(root.children);
+      root.children.clear();
+      root.children.addAll(rolledUp);
+
+      // need to promote a child if it's an only child
+      if (root.children.size() == 1) {
+        return root.children.get(0);
+      }
+    }
+
+    return root;
+  }
+
+  /*
+   * Walks an expression's AST and appends a string representation to a supplied StringBuilder. This
+   * method adds parens where necessary.
+   */
+  public static void stringify(Node root, byte[] expression, StringBuilder out) {
```

Comment:
Is this still needed?

Reply:
yeah it used by a public normalize method.  Looking at the code that method did not have an `@Override` tag so I added that.